pkg/plugins/policies/meshtrafficpermission/graph/reachable_services_graph.go

```diff
@@ -0,0 +1,161 @@
+	return BuildGraph(services, mtps.Items)
+}
+
+func BuildServices(
+	dataplanes []*mesh.DataplaneResource,
+	externalServices []*mesh.ExternalServiceResource,
+) map[string]mesh_proto.SingleValueTagSet {
+	services := map[string]mesh_proto.SingleValueTagSet{}
+	for _, dp := range dataplanes {
+		set := dp.Spec.TagSet()
+		for _, svc := range set.Values(mesh_proto.ServiceTag) {
+			if _, ok := services[svc]; ok {
+				continue
+			}
+			services[svc] = map[string]string{}
+			for tag := range SupportedTags {
+				if values := set.Values(tag); len(values) > 0 {
+					services[svc][tag] = values[0]
+				}
+			}
+		}
+	}
+	for _, es := range externalServices {
+		services[es.Spec.Tags[mesh_proto.ServiceTag]] = map[string]string{}
+	}
+	return services
+}
+
+func BuildGraph(services map[string]mesh_proto.SingleValueTagSet, mtps []*mtp_api.MeshTrafficPermissionResource) *Graph {
+	resources := context.Resources{
+		MeshLocalResources: map[core_model.ResourceType]core_model.ResourceList{
+			mtp_api.MeshTrafficPermissionType: &mtp_api.MeshTrafficPermissionResourceList{
+				Items: replaceTopLevelSubsets(mtps),
+			},
+		},
+	}
+
+	graph := NewGraph()
+
+	for service, tags := range services {
+		// build artificial dpp for matching
+		dp := mesh.NewDataplaneResource()
+		dpTags := maps.Clone(tags)
+		dpTags[mesh_proto.ServiceTag] = service
+		dp.Spec = &mesh_proto.Dataplane{
+			Networking: &mesh_proto.Dataplane_Networking{
+				Address: "1.1.1.1",
+				Inbound: []*mesh_proto.Dataplane_Networking_Inbound{
+					{
+						Tags: dpTags,
+						Port: 1234,
+					},
+				},
+			},
+		}
+
+		matched, err := matchers.MatchedPolicies(mtp_api.MeshTrafficPermissionType, dp, resources)
+		if err != nil {
+			log.Error(err, "service could not be matched. It won't be reached by any other service", "service", service)
+			continue // it's better to ignore one service that to break the whole graph
+		}
+
+		rl, ok := matched.FromRules.Rules[core_rules.InboundListener{
+			Address: "1.1.1.1",
+			Port:    1234,
+		}]
+		if !ok {
+			continue
+		}
+
+		graph.rules[service] = rl
+	}
+
+	return graph
+}
+
+// replaceTopLevelSubsets replaces subsets present in top-level target ref.
+// Because we need to do policy matching on services instead of individual proxies, we have to handle subsets in a special way.
+// What we do is we only support subsets with predefined tags listed in SupportedTags.
+// This assumes that tags listed in SupportedTags are stable between all instances of a given service.
+// Otherwise, we fallback to higher level target ref (MeshSubset -> Mesh, MeshServiceSubset -> MeshService).
+//
+// Alternatively, we could have computed all common tags between instances of a given service and then allow subsets with those common tags.
+// However, this would require calling this function for every service.
+func replaceTopLevelSubsets(mtps []*mtp_api.MeshTrafficPermissionResource) []*mtp_api.MeshTrafficPermissionResource {
```

Comment:
Wait, but this transformation of policies can change the final configuration, i.e given:

```yaml
name: aaa
targetRef:
  kind: MeshSubset
  tags:
    k8s.kuma.io/namespace: ns1
from: 
  - targetRef:
      kind: MeshService
      name: backend
    default:
      action: Allow
---
name: bbb
targetRef:
  kind: MeshSubset
  tags:
    non-supported-tag: value
from: 
  - targetRef:
      kind: MeshService
      name: backend
    default:
      action: Deny
```
since kinds are equal and `bbb > aaa` then the final result is `Deny` for `backend`. After `replaceTopLevelSubsets` it will look like this:

```yaml
name: aaa
targetRef:
  kind: MeshSubset
  tags:
    k8s.kuma.io/namespace: ns1
from: 
  - targetRef:
      kind: MeshService
      name: backend
    default:
      action: Allow
---
name: bbb
targetRef:
  kind: Mesh
from: 
  - targetRef:
      kind: MeshService
      name: backend
    default:
      action: Deny
```
now MeshSubset is more specific than Mesh so the final result is `Allow` for `backend`.

Reply:
should we delete all froms with Deny whenever we replace subset?

It's what I had before when I replaced subsets in "form" I ignored the one with deny

Reply:
I don't think it will work, just imagine the example I gave but swap `Allow` and `Deny`:
```yaml
name: aaa
targetRef:
  kind: MeshSubset
  tags:
    k8s.kuma.io/namespace: ns1
from: 
  - targetRef:
      kind: MeshService
      name: backend
    default:
      action: Deny
---
name: bbb
targetRef:
  kind: MeshSubset
  tags:
    non-supported-tag: value
from: 
  - targetRef:
      kind: MeshService
      name: backend
    default:
      action: Allow
```
after replacing it will be:
```yaml
name: aaa
targetRef:
  kind: MeshSubset
  tags:
    k8s.kuma.io/namespace: ns1
from: 
  - targetRef:
      kind: MeshService
      name: backend
    default:
      action: Deny
---
name: bbb
targetRef:
  kind: Mesh
from: 
  - targetRef:
      kind: MeshService
      name: backend
    default:
      action: Allow
```

Reply:
Why it wouldn't work in this case? Yes, the graph would create clusters for more clients than needed in a perfect scenario, but it wouldn't block the traffic, right?

Reply:
before you had `backend: Allow`, after replacement, you have `backend: Deny`, so `backend` won't receive the outbound

Reply:
talked offline that the solution is to keep existing target ref but get rid of unsupported tags